ipam/http.go

```diff
@@ -16,6 +16,43 @@ func badRequest(w http.ResponseWriter, err error) {
 	common.Log.Warningln("[allocator]:", err.Error())
 }
 
+func extractCIDR(w http.ResponseWriter, vars map[string]string) (address.CIDR, bool) {
+	cidrStr := vars["ip"] + "/" + vars["prefixlen"]
+	subnetAddr, cidr, err := address.ParseCIDR(cidrStr)
+	if err != nil {
+		badRequest(w, err)
+		return address.CIDR{}, false
+	}
+	if cidr.Start != subnetAddr {
+		badRequest(w, fmt.Errorf("Invalid subnet %s - bits after network prefix are not all zero", cidrStr))
+		return address.CIDR{}, false
+	}
+	return cidr, true
+}
+
+func doAllocate(alloc *Allocator, dockerCli *docker.Client, w http.ResponseWriter, r *http.Request, vars map[string]string, subnet address.CIDR) {
+	closedChan := w.(http.CloseNotifier).CloseNotify()
+	ident := vars["id"]
+	addr, err := alloc.Allocate(ident, subnet.HostRange(), closedChan)
+	if err != nil {
+		if _, ok := err.(*errorCancelled); ok { // cancellation is not really an error
+			common.Log.Infoln("[allocator]:", err.Error())
+			fmt.Fprint(w, "cancelled")
+			return
+		}
+		badRequest(w, err)
+		return
+	}
+	if r.FormValue("check-alive") == "true" && dockerCli != nil && dockerCli.IsContainerNotRunning(ident) {
+		common.Log.Infof("[allocator] '%s' is not running: freeing %s", ident, addr)
+		alloc.Free(ident, addr)
+		fmt.Fprint(w, "cancelled")
```

Comment:
Note this is a case not called out by #1418, but it seems reasonable to do the same thing - previously it would return blank which would cause an error elsewhere in the weave script.